common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,379 @@
+        //When in Usb Discovery mode, Yubikeys that plug into the device will be accessible
+        // once the user provides permission via the Android permission dialog.
         mYubiKitManager = new YubiKitManager(mActivity.getApplicationContext());
         mYubiKitManager.startUsbDiscovery(new UsbConfiguration(), new Callback<UsbYubiKeyDevice>() {
             @Override
             public void invoke(UsbYubiKeyDevice device) {
                 Logger.info(TAG, "A YubiKey device was connected");
-                device.setOnClosed(new Runnable() {
-                    @Override
-                    public void run() {
-                        Logger.info(TAG, "A YubiKey device was disconnected");
-                    }
-                });
+                synchronized (deviceLock) {
+                    //set device
+                    mDevice = device;
+                    mDevice.setOnClosed(new Runnable() {
+                        @Override
+                        public void run() {
+                            synchronized (deviceLock) {
+                                Logger.info(TAG, "A YubiKey device was disconnected");
+                                mDevice = null;
+                                //Call onCancel on current dialog, if dialog is showing.
+                                synchronized (smartcardDialogLock) {
+                                    if (mCurrentDialog != null) {
+                                        mCurrentDialog.onCancelCba();
+                                        //Reset dialog to null
+                                        mCurrentDialog = null;
+                                    }
+                                }
+                            }
+                        }
+                    });
+                }
             }
         });
     }
 
+    //Called when a ClientCertRequest is received by the WebViewClient.
+    //Offers the user to choose a certificate to authenticate with based on whether or not a YubiKey is plugged in and has permission to be connected.
     @TargetApi(Build.VERSION_CODES.LOLLIPOP)
     @Override
     public Void processChallenge(@NonNull final ClientCertRequest request) {
-        //final String methodTag = TAG + ":processChallenge";
+        final String methodTag = TAG + ":processChallenge";
+        // Check if a YubiKey device has been discovered and has permission to be connected.
+        synchronized (deviceLock) {
+            if (mDevice != null) {
+                //A connection to the YubiKey needs to be made in order to read the certificates off it.
+                mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+                    @Override
+                    public void invoke(Result<UsbSmartCardConnection, IOException> value) {
+                        try {
+                            final SmartCardConnection c = value.getValue();
```

Comment:
what if value returns an exception?

Reply:
The method itself seems to handle throwing an exception. When `getValue()` is called, an IOException is thrown if the value is null. 
https://github.com/Yubico/yubikit-android/blob/0f04eeb3c9c9fc06eca29883b4ef2fa90293ffad/core/src/main/java/com/yubico/yubikit/core/util/Result.java#L35
But let me know if I'm misunderstanding this.